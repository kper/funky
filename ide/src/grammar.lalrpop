use std::str::FromStr;
use crate::ir::ast::*;

grammar;

Comma<T>: Vec<T> = { 
    <v:(<T> ",")*> <e:T?> => match e {
        None => v,
        Some(e) => {
            let mut v = v;
            v.push(e);
            v
        }
    }
};

pub Program : Program = {
   <functions:Function*> => Program {
       functions: functions
   } 
};

pub Function : Function = {
    "define" <name:Id> "(result" <result:Num> ")" "(define" <def:Reg*> ")" "{" <instr:Instruction*>  "}" ";" => {
        Function {
            name,
            params: Vec::new(),
            definitions: def,
            results_len: result as usize,
            instructions: instr
        }
    },
    "define" <name:Id> "(param" <reg:Reg*> ")" "(result" <result:Num> ")" "(define" <def:Reg*> ")" "{" <instr:Instruction*>  "}" ";" => {
        Function {
            name,
            params: reg,
            definitions: def,
            results_len: result as usize,
            instructions: instr
        }
    },
    "define" <name:Name> "(result" <result:Num> ")" "(define" <def:Reg*> ")" "{" <instr:Instruction*>  "}" ";" => {
        Function {
            name,
            params: Vec::new(),
            definitions: def,
            results_len: result as usize,
            instructions: instr
        }
    },
    "define" <name:Name> "(param" <reg:Reg*> ")" "(result" <result:Num> ")" "(define" <def:Reg*> ")" "{" <instr:Instruction*>  "}" ";" => {
        Function {
            name,
            params: reg,
            definitions: def,
            results_len: result as usize,
            instructions: instr
        }
    }
};

pub Instruction : Instruction = {
    "BLOCK" <n:Id> => Instruction::Block(n),
    <dest:Reg> "=" <src:Reg> => Instruction::Assign(dest, src),
    <dest:Reg> "=" <n:Num> => Instruction::Const(dest, n),
    <dest:Reg> "=" "UNKNOWN" => Instruction::Unknown(dest),
    <dest:Reg> "=" "op" <src:Reg> => Instruction::Unop(dest, src),
    <dest:Reg> "=" <src1:Reg> "op" <src2:Reg> => Instruction::BinOp(dest, src1, src2),
    <dest:Reg> "=" "phi" <src1:Reg> <src2:Reg> => Instruction::Phi(dest, src1, src2),
    <dest:Reg> "=" "STORE" <src:Reg> "TO" <offset:Reg> "ALIGN" <align:Num> <size:Num> => Instruction::Store,
    "GOTO" <n:Id> => Instruction::Jump(n),
    "IF" <reg:Reg> "THEN" "GOTO" <n:Id> "ELSE" "GOTO" <n2:Id> => {
        Instruction::Conditional(reg, vec![n, n2])
    },
    "IF" <reg:Reg> "THEN" "GOTO" <n:Id> => {
        Instruction::Conditional(reg, vec![n])
    },
    "TABLE GOTO" <n:Id*> "ELSE" "GOTO" <n2:Id> => {
        let mut n = n.clone();
        n.push(n2);
        Instruction::Table(n)
    },
    "KILL" <reg:Reg> => Instruction::Kill(reg),
    "CALL" <n:Name> "(" <r:Reg*> ")" => Instruction::Call(n, r, vec![]),
    <dest:Reg*> "<-" "CALL" <n:Name> "(" <r:Reg*> ")" => Instruction::Call(n, r, dest),
    "CALL" <n:Id> "(" <r:Reg*> ")" => Instruction::Call(n, r, vec![]),
    <dest:Reg*> "<-" "CALL" <n:Id> "(" <r:Reg*> ")" => Instruction::Call(n, r, dest),
    "RETURN" <n:Reg*> ";" => {
        Instruction::Return(n)
    },
    "CALL INDIRECT" <n:Id*> "(" <p:Reg*> ")" => Instruction::CallIndirect(n, p, vec![]),
    <dest:Reg*> "<-" "CALL INDIRECT" <n:Id*> "(" <p:Reg*> ")" => Instruction::CallIndirect(n, p, dest),
};

Id : String = {
    <n:Num> => (n as usize).to_string()
};

pub Reg: String = {
    "%" <n:Num> => format!("%{}", n),
    "%" <n:Name> => format!("%{}", n),
};

pub Num: f64 = <s:r"[+-]?[0-9]*[.,]?[0-9]+"> => f64::from_str(s).unwrap();
pub Name: String = <s:r"[a-zA-Z]+"> => s.to_string();