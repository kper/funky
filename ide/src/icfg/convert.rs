/// This module is responsible to parse
/// the webassembly AST to a graph

use crate::counter::Counter;
use crate::icfg::graph::Graph;
use crate::ssa::ast::Instruction;
use anyhow::{bail, Context, Result};

use log::debug;

use crate::icfg::graph::Fact;
use std::collections::HashMap;

use crate::ssa::ast::Program;

#[derive(Debug)]
pub struct Convert {
    block_counter: Counter,
    functions: HashMap<String, usize>, // function name to fact id
    registration_returns: HashMap<String, (Vec<Fact>, Vec<String>)>, // function name to (all facts + dest regs)
}

impl Convert {
    pub fn new() -> Self {
        Self {
            block_counter: Counter::default(),
            functions: HashMap::new(),
            registration_returns: HashMap::new(),
        }
    }

    pub fn visit(&mut self, prog: Program) -> Result<Graph> {
        debug!("Convert intermediate repr to graph");

        let mut graph = Graph::new();

        for function in prog.functions.iter() {
            debug!("Init function {}", function.name);
            graph.init_function(function);
            //graph.add_row(&function.name, format!("init {}", function.name), &mut Vec::new())?;
        }

        //debug!("graph {:#?}", graph);

        for function in prog.functions.iter() {
            debug!("Creating graph from function {}", function.name);

            // New function
            graph.epoch.push();

            let mut iterator =
                InstructionIterator::new(function.instructions.iter().collect::<Vec<_>>());

            for instruction in &mut iterator {
                match instruction {
                    Instruction::Const(reg, _val) => {
                        debug!("Adding const");

                        graph.add_row(
                            &function.name,
                            format!("before {:?}", instruction),
                        )?;

                        graph.add_var(&function.name, &reg)?;

                        graph.add_row(
                            &function.name,
                            format!("after {:?}", instruction),
                        )?;
                    }
                    Instruction::Assign(dest, src) => {
                        debug!("Assignment");

                        graph.add_row(
                            &function.name,
                            format!("before {:?}", instruction),
                        )?;

                        graph.add_assignment(&function.name, &dest, &src)?;

                        graph.add_row(
                            &function.name,
                            format!("after {:?}", instruction),
                        )?;
                    }
                    Instruction::Unop(dest, src) => {
                        debug!("Unop");
                        graph.add_row(
                            &function.name,
                            format!("before {:?}", instruction),
                        )?;
                        graph.add_unop(&function.name, &dest, &src)?;
                        graph.add_row(
                            &function.name,
                            format!("after {:?}", instruction),
                        )?;
                    }
                    Instruction::BinOp(dest, src1, src2) => {
                        debug!("Binop");
                        graph.add_row(
                            &function.name,
                            format!("before {:?}", instruction),
                        )?;

                        graph.add_binop(&function.name, &dest, &src1, &src2)?;
                        graph.add_row(
                            &function.name,
                            format!("after {:?}", instruction),
                        )?;
                    }
                    Instruction::Kill(dest) => {
                        debug!("Kill");
                        graph.add_row(
                            &function.name,
                            format!("before {:?}", instruction),
                        )?;
                        graph.kill_var(&function.name, &dest)?;
                        graph.add_row(
                            &function.name,
                            format!("after {:?}", instruction),
                        )?;
                    }
                    Instruction::Call(name, _params, dest_regs) => {
                        graph.add_empty_vars(&function.name, &dest_regs)?;
                        graph.add_row(
                            &function.name,
                            format!("before {:?}", instruction),
                        )?;
                        let meeting_facts = graph.add_call_to_return(
                            &function.name,
                            format!("Return from {}", name),
                        )?;

                        // Expect a return edge from `name` with function.name to the meeting facts
                        self.registration_returns
                            .insert(name.clone(), (meeting_facts, dest_regs.clone()));
                    }
                    _ => {}
                }
            }
            
            // function ended
            graph.epoch.pop();
        }

        for function in prog.functions.iter() {
            let mut iterator =
                InstructionIterator::new(function.instructions.iter().collect::<Vec<_>>());

            for instruction in &mut iterator {
                match instruction {
                    Instruction::Call(name, params, _dest_regs) => {
                        debug!("Call {}", name);

                        let lookup_function = prog
                            .functions
                            .iter()
                            .find(|x| &x.name == name)
                            .context("Function not found")?;

                        graph.add_call(&function.name, &lookup_function, name, params)?;
                    }
                    _ => {}
                }
            }
        }

        for (goal_function, (meeting_facts, dest_regs)) in self.registration_returns.drain() {
            if let Some((goal_function_name, goal_function_results_len, goal_function_last_facts)) = 
                graph.functions.get(&goal_function).map(|x| (&x.name, x.results_len, x.last_facts.clone())) {

                if goal_function_results_len != dest_regs.len() {
                    bail!("Mismatch results with call of {}.\nExpected results: {}\nActual results: {}", 
                        goal_function_name, 
                        goal_function_results_len, 
                        dest_regs.len());
                }

                graph.add_return(
                    &goal_function_last_facts,
                    meeting_facts,
                    &dest_regs,
                )?;
            }
        }

        return Ok(graph);
    }
}

/// Highlevel iterator, that has the
/// ability to jump to blocks
struct InstructionIterator<'a> {
    instructions: Vec<&'a Instruction>,
    current: usize,
    blocks: HashMap<&'a String, usize>, // BlockId to index in instructions
}

impl<'a> InstructionIterator<'a> {
    pub fn new(instructions: Vec<&'a Instruction>) -> Self {
        let mut v = Self {
            instructions,
            current: 0,
            blocks: HashMap::default(),
        };

        v.create_block_jump_list();

        v
    }

    fn create_block_jump_list(&mut self) {
        for (index, block) in self
            .instructions
            .iter()
            .enumerate()
            .filter(|(_id, x)| matches!(x, Instruction::Block(_)))
        {
            match block {
                Instruction::Block(id) => {
                    self.blocks.insert(id, index);
                }
                _ => {
                    panic!("Only expecting blocks");
                }
            }
        }
    }

    pub fn jump_to(&mut self, block_id: &String) -> Result<()> {
        debug!("Jump to block {}", block_id);
        if let Some(index) = self.blocks.get(block_id) {
            self.current = *index;
            return Ok(());
        } else {
            bail!("Block {} does not exist", block_id);
        }
    }

    #[allow(dead_code)]
    pub fn peek(&self) -> Option<&'a Instruction> {
        self.instructions.get(self.current).map(|x| *x)
    }
}

impl<'a> std::iter::Iterator for &mut InstructionIterator<'a> {
    type Item = &'a Instruction;

    fn next(&mut self) -> Option<Self::Item> {
        let item = self.instructions.get(self.current);
        self.current += 1;
        let item = item.map(|x| *x);

        match item {
            Some(Instruction::Block(_)) => {
                return self.next();
            }
            _ => {}
        }

        if let Some(&Instruction::Jump(ref id)) = item {
            if self.jump_to(id).is_ok() {
                self.next()
            } else {
                debug!("Block not found, therefore ending");
                None
            }
        } else {
            item
        }
    }
}
