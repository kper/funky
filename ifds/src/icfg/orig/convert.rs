/// This module is responsible to parse
/// the webassembly AST to a graph
use crate::icfg::graph::*;
use crate::icfg::state::State;
use crate::ir::ast::Function as AstFunction;
use crate::ir::ast::Instruction;

use crate::solver::Request;
use anyhow::{Context, Result};
use std::collections::VecDeque;

use log::debug;

use crate::ir::ast::Program;

use crate::icfg::naive::convert::Convert as NaiveConvert;
use std::collections::HashMap;

type CallerFunction = String;
type PC = usize;
type Function = String;
type CallResolver = HashMap<Function, Vec<(CallerFunction, PC, Vec<String>)>>;
pub(crate) struct Ctx<'a> {
    pub graph: &'a mut Graph,
    pub new_graph: &'a mut Graph,
    pub state: &'a mut State,
    pub call_resolver: &'a CallResolver,
}

/// Central datastructure for the computation of the IFDS problem.
#[derive(Debug, Default)]
pub struct OriginalConvert;

impl OriginalConvert {
    /// Computes a graph by a given program and `req` ([`Request`]).
    /// The `variable` in `req` doesn't matter. It only matters the `function` and `pc`.
    pub fn visit(&mut self, prog: &Program, req: &Request) -> Result<(Graph, State)> {
        let mut fact_gen = NaiveConvert::default();
        let (mut graph, mut state, call_resolver) = fact_gen
            .visit(prog)
            .context("Naive fact generation failed")?;

        let mut new_graph = Graph::default();

        let mut ctx = Ctx {
            graph: &mut graph,
            state: &mut state,
            new_graph: &mut new_graph,
            call_resolver: &call_resolver,
        };

        self.tabulate(&mut ctx, prog, req)?;

        Ok((new_graph, state))
    }

    fn tabulate<'a>(&mut self, ctx: &mut Ctx<'a>, prog: &Program, req: &Request) -> Result<()> {
        debug!("Convert intermediate repr to graph");

        let function = prog
            .functions
            .iter()
            .find(|x| x.name == req.function)
            .context("Cannot find function")?;

        let facts = ctx
            .state
            .get_facts_at(&function.name, req.pc)?
            .cloned()
            .collect::<Vec<_>>();

        let mut path_edge = Vec::new();
        let mut worklist = VecDeque::new();
        let mut summary_edge = Vec::new();
        let mut normal_flows_debug = Vec::new();

        let init = facts.get(0).unwrap().clone();

        // self loop for taut
        self.propagate(
            &mut ctx.new_graph,
            &mut path_edge,
            &mut worklist,
            Edge::Path {
                from: init.clone(),
                to: init.clone(),
            },
        )?;

        self.forward(
            &prog,
            &function,
            &mut path_edge,
            &mut worklist,
            &mut summary_edge,
            &mut normal_flows_debug,
            ctx,
            req.pc,
        )?;

        Ok(())
    }

    fn propagate(
        &self,
        graph: &mut Graph,
        path_edge: &mut Vec<Edge>,
        worklist: &mut VecDeque<Edge>,
        e: Edge,
    ) -> Result<()> {
        let from = e.get_from();
        let to = e.to();

        let f = path_edge
            .iter()
            .find(|x| x.get_from() == from && x.to() == to);

        if f.is_none() {
            debug!("Propagate {:#?}", e);
            graph.edges.push(e.clone());
            path_edge.push(e.clone());
            worklist.push_back(e);
        }

        Ok(())
    }

    fn forward<'a>(
        &mut self,
        program: &Program,
        _function: &AstFunction,
        path_edge: &mut Vec<Edge>,
        worklist: &mut VecDeque<Edge>,
        summary: &mut Vec<Edge>,
        normal_flows_debug: &mut Vec<Edge>,
        ctx: &mut Ctx<'a>,
        start_pc: usize,
    ) -> Result<()> {
        while let Some(edge) = worklist.pop_front() {
            debug!("Popping edge from worklist {:#?}", edge);

            assert!(
                matches!(edge, Edge::Path { .. }),
                "Edge in the worklist has wrong type"
            );

            let d1 = edge.get_from();
            let d2 = edge.to();

            let pc = edge.to().next_pc;
            debug!("Instruction pointer is {}", pc);

            let instructions = &program
                .functions
                .iter()
                .find(|x| x.name == d2.function)
                .context("Cannot find function")?
                .instructions;
            let n = instructions.get(pc);
            debug!("=> Instruction {:?}", n);

            if let Some(n) = n {
                match n {
                    Instruction::Call(callee, _params, _dest) => {
                        // handle_call

                        let caller_function = program
                            .functions
                            .iter()
                            .find(|x| x.name == d1.function)
                            .unwrap();

                        let callee_function = program
                            .functions
                            .iter()
                            .find(|x| &x.name == callee)
                            .unwrap();

                        let flow_edges = ctx.graph.edges.iter().filter(|x| {
                            x.get_from().function == caller_function.name
                                && x.get_from().belongs_to_var == d2.belongs_to_var
                                && x.get_from().next_pc == d2.next_pc
                                && x.to().function == callee_function.name
                                && x.to().next_pc == 0
                        });

                        for f in flow_edges.into_iter() {
                            let to = f.to();
                            self.propagate(
                                &mut ctx.new_graph,
                                path_edge,
                                worklist,
                                Edge::Path {
                                    from: to.clone(),
                                    to: to.clone(),
                                },
                            )?;
                        }

                        let flow_edges = ctx
                            .graph
                            .edges
                            .iter()
                            .chain(summary.iter())
                            .filter(|x| {
                                x.get_from().function == caller_function.name
                                    && x.get_from().belongs_to_var == d2.belongs_to_var
                                    && x.get_from().next_pc == d2.next_pc
                                    && x.to().function == caller_function.name
                                    //&& x.to().belongs_to_var == x.get_from().belongs_to_var
                                    && x.to().next_pc == x.get_from().next_pc + 1
                            })
                            .map(|x| x.to());

                        for to in flow_edges.into_iter() {
                            assert_eq!(d1.function, to.function);

                            self.propagate(
                                &mut ctx.new_graph,
                                path_edge,
                                worklist,
                                Edge::Path {
                                    from: d1.clone(),
                                    to: to.clone(),
                                },
                            )?;
                        }
                    }
                    _ => {
                        let new_function = program
                            .functions
                            .iter()
                            .find(|x| x.name == d2.function)
                            .unwrap();

                        let flow_edges =
                            ctx.graph
                                .edges
                                .iter()
                                .filter(|x| x.is_normal())
                                .filter(|x| {
                                    x.get_from().function == new_function.name
                                        && x.get_from().belongs_to_var == d2.belongs_to_var
                                        && x.get_from().next_pc == d2.next_pc
                                });

                        for f in flow_edges.into_iter() {
                            debug!("Normal flow {:#?}", f);
                            let to = f.to();

                            normal_flows_debug.push(f.clone());

                            self.propagate(
                                &mut ctx.new_graph,
                                path_edge,
                                worklist,
                                Edge::Path {
                                    from: d1.clone(),
                                    to: to.clone(),
                                },
                            )?;
                        }
                    }
                }
            } else {
                // end procedure

                if let Some(callers) = ctx.call_resolver.get(&d1.function) {
                    let callers = callers
                        .iter()
                        .map(|(caller, _pc, _dests)| (caller, &d1.function));

                    for (caller, callee) in callers {
                        assert_eq!(callee, &d1.function);

                        let d4 = ctx
                            .graph
                            .edges
                            .iter()
                            .filter(|x| x.is_call())
                            .filter(|x| {
                                &x.get_from().function == caller
                                    && x.to().function == d1.function
                                    && x.to().belongs_to_var == d1.belongs_to_var
                                    && x.to().next_pc == d1.next_pc
                            })
                            .map(|x| x.get_from());

                        for d4 in d4.into_iter() {
                            let d5 = ctx
                                .graph
                                .edges
                                .iter()
                                .filter(|x| x.is_return())
                                .filter(|x| {
                                    x.get_from().belongs_to_var == d2.belongs_to_var
                                    && x.get_from().function == d2.function
                                    && x.get_from().next_pc == d2.next_pc - 1 //because applied before
                                    && &x.to().function == &d4.function
                                    // && x.to().next_pc == pc + 1
                                })
                                .map(|x| x.to());

                            for d5 in d5 {
                                if summary
                                    .iter()
                                    .find(|x| {
                                        &x.get_from().function == caller
                                            && x.get_from().belongs_to_var == d4.belongs_to_var
                                            && &x.to().function == caller
                                            && x.to().next_pc == x.get_from().next_pc + 1
                                            && x.to().belongs_to_var == d5.belongs_to_var
                                    })
                                    .is_none()
                                {
                                    let ret = d4.clone();
                                    summary.push(Edge::Summary {
                                        from: d4.clone(),
                                        to: ret.clone(),
                                    });

                                    let edges = ctx
                                        .graph
                                        .edges
                                        .iter()
                                        .filter(|x| {
                                            &x.get_from().function == caller
                                                && &x.to().function == caller
                                                && x.get_from().next_pc == start_pc
                                                && x.to().belongs_to_var == d4.belongs_to_var
                                                && x.to().next_pc == d4.next_pc
                                        })
                                        .map(|x| x.get_from());

                                    for d3 in edges {
                                        let mut ret = ret.clone();
                                        ret.belongs_to_var = d5.belongs_to_var.clone();

                                        let init = ctx
                                            .state
                                            .get_facts_at(caller, 0)?
                                            .find(|x| x.belongs_to_var == d3.belongs_to_var)
                                            .context("Cannot find fact")?
                                            .clone();

                                        self.propagate(
                                            &mut ctx.new_graph,
                                            path_edge,
                                            worklist,
                                            Edge::Path {
                                                from: init,
                                                to: ret,
                                            },
                                        )?;
                                    }
                                }
                            }
                        }
                    }
                }
            }
        }

        //graph.edges.extend_from_slice(&path_edge);
        //graph.edges.extend_from_slice(&normal_flows_debug);
        //graph.edges.extend_from_slice(&summary_edge);

        Ok(())
    }
}
