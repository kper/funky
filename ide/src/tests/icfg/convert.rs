use crate::icfg::convert::Convert;
//use crate::icfg::graphviz::render_to;
use insta::assert_snapshot;
use std::io::Cursor;

use crate::grammar::*;

macro_rules! ir {
    ($name:expr, $ir:expr) => {
        let mut convert = Convert::new();

        let prog = ProgramParser::new().parse(&$ir).unwrap();

        let res = convert.visit(prog).unwrap();

        //let mut dot = Cursor::new(Vec::new());
        //render_to(&res, &mut dot);

        /* 
        assert_snapshot!(
            format!("{}_dot", $name),
            std::str::from_utf8(dot.get_ref()).unwrap()
        );*/
    };
}

#[test]
fn test_ir_const() {
    ir!(
        "test_ir_const",
        "
         define test (result 0) {
            %0 = 1
         };
    "
    );
}

#[test]
fn test_ir_double_const() {
    ir!(
        "test_ir_double_const",
        "
         define test (result 0) {
            %0 = 1
            %1 = 1
         };
    "
    );
}

#[test]
fn test_ir_assignment() {
    ir!(
        "test_ir_assignment",
        "
         define test (result 0) {
            %1 = 1
            %0 = %1
         };
    "
    );
}

#[test]
fn test_ir_double_assignment() {
    ir!(
        "test_ir_double_assignment",
        "
         define test (result 0) {
            %1 = 1
            %0 = %1
            %0 = %1
         };
    "
    );
}

#[test]
fn test_ir_block() {
    ir!(
        "test_ir_block",
        "define test (result 0) {
            BLOCK 0
            %0 = 1
            GOTO 1
            BLOCK 1
            %1 = 2
        };"
    );
}

#[test]
fn test_ir_killing() {
    ir!(
        "test_ir_killing",
        "define test (result 0) {
            %0 = 1
            %0 = 2
        };"
    );
}

#[test]
fn test_ir_unop() {
    ir!(
        "test_ir_unop",
        "define test (result 0)  {
            %0 = 1
            %1 = op %0
            %1 = op %0   
        };"
    );
}

#[test]
fn test_ir_binop() {
    ir!(
        "test_ir_binop",
        "define test (result 0)  {
            %0 = 1
            %1 = 1
            %2 = %0 op %1
            %2 = %1 op %0   
        };"
    );
}

#[test]
fn test_ir_killing_op() {
    ir!(
        "test_ir_killing_op",
        "define test (result 0)  {
            %0 = 1
            %1 = 1
            KILL %0
            KILL %1
            %2 = 1
        };"
    );
}

#[test]
fn test_ir_functions() {
    ir!(
        "test_ir_functions",
        "define test (result 0) {
            %0 = 1
            CALL mytest(%0)
        };
        define mytest (param %0) (result 0)  {
            %0 = 2   
            %1 = 3
        };"
    );
}

#[test]
fn test_ir_return_values() {
    ir!(
        "test_ir_return_values",
        "define test (result 0) {
            %0 = 1
            %1 <- CALL mytest(%0)
        };
        define mytest (param %0) (result 1) {
            %0 = 2   
            %1 = 3
        };"
    );
}

#[should_panic]
#[test]
fn test_ir_return_mismatched_values() {
    // Assigning %1 but `result 0`
    ir!(
        "test_ir_mismatched_functions",
        "define test (result 0) {
            %0 = 1
            %1 <- CALL mytest(%0)
        };
        define mytest (param %0) (result 0) {
            %0 = 2   
            %1 = 3
        };"
    );
}

#[should_panic]
#[test]
fn test_ir_return_mismatched_values2() {
    // Assigning void but `result 1`
    ir!(
        "test_ir_mismatched_functions",
        "define test (result 0) {
            %0 = 1
            CALL mytest(%0)
        };
        define mytest (param %0) (result 1) {
            %0 = 2   
            %1 = 3
        };"
    );
}
