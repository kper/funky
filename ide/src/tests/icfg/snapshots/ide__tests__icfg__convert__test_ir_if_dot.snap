---
source: src/tests/icfg/convert.rs
expression: output

---
   \documentclass{standalone}
            \usepackage{pgf, tikz}
            \usetikzlibrary{arrows, automata}
            \begin{document}
                \begin{tikzpicture}
                    [node distance=3cm]
                    \tikzstyle{every state}=[
                        draw = black,
                        thick,
                        fill = white,
                        minimum size = 4mm
                    ]

                    \node[circle,fill,inner sep=1pt,label=left:taut] (0) at (0, 0) {};
\node[circle,fill,inner sep=1pt,label=left:\%0] (1) at (1, 0) {};
\node[circle,fill,inner sep=1pt,label=left:\%1] (2) at (2, 0) {};
\node[circle,fill,inner sep=1pt,label=left:\%2] (3) at (3, 0) {};
\node[circle,fill,inner sep=1pt,label=left:taut] (4) at (0, 1) {};
\node[circle,fill,inner sep=1pt,label=left:\%0] (5) at (1, 1) {};
\node[circle,fill,inner sep=1pt,label=left:\%1] (6) at (2, 1) {};
\node[circle,fill,inner sep=1pt,label=left:\%2] (7) at (3, 1) {};
\node[circle,fill,inner sep=1pt,label=left:taut] (8) at (0, 2) {};
\node[circle,fill,inner sep=1pt,label=left:\%0] (9) at (1, 2) {};
\node[circle,fill,inner sep=1pt,label=left:\%1] (10) at (2, 2) {};
\node[circle,fill,inner sep=1pt,label=left:\%2] (11) at (3, 2) {};
\node[circle,fill,inner sep=1pt,label=left:taut] (12) at (0, 3) {};
\node[circle,fill,inner sep=1pt,label=left:\%0] (13) at (1, 3) {};
\node[circle,fill,inner sep=1pt,label=left:\%1] (14) at (2, 3) {};
\node[circle,fill,inner sep=1pt,label=left:\%2] (15) at (3, 3) {};
\node[circle,fill,inner sep=1pt,label=left:taut] (16) at (0, 4) {};
\node[circle,fill,inner sep=1pt,label=left:\%0] (17) at (1, 4) {};
\node[circle,fill,inner sep=1pt,label=left:\%1] (18) at (2, 4) {};
\node[circle,fill,inner sep=1pt,label=left:\%2] (19) at (3, 4) {};
\node[circle,fill,inner sep=1pt,label=left:taut] (20) at (0, 5) {};
\node[circle,fill,inner sep=1pt,label=left:\%0] (21) at (1, 5) {};
\node[circle,fill,inner sep=1pt,label=left:\%1] (22) at (2, 5) {};
\node[circle,fill,inner sep=1pt,label=left:\%2] (23) at (3, 5) {};
\node[circle,fill,inner sep=1pt,label=left:taut] (24) at (0, 6) {};
\node[circle,fill,inner sep=1pt,label=left:\%0] (25) at (1, 6) {};
\node[circle,fill,inner sep=1pt,label=left:\%1] (26) at (2, 6) {};
\node[circle,fill,inner sep=1pt,label=left:\%2] (27) at (3, 6) {};
\node[font=\tiny] (note_0) at (-1.5, 0.5) {Block(0)};
\node[font=\tiny] (note_1) at (-1.5, 1.5) {Const(0, 1.0)};
\node[font=\tiny] (note_2) at (-1.5, 2.5) {Conditional(1, true)};
\node[font=\tiny] (note_3) at (-1.5, 3.5) {Jump(0)};
\node[font=\tiny] (note_4) at (-1.5, 4.5) {Const(1, 2.0)};
\node[font=\tiny] (note_5) at (-1.5, 5.5) {Const(2, 3.0)};
		\path[->] (0) edge (4);
		\path[->] (1) edge (5);
		\path[->] (2) edge (6);
		\path[->] (3) edge (7);
		\path[->] (4) edge (9);
		\path[->] (4) edge (8);
		\path[->] (6) edge (10);
		\path[->] (7) edge (11);
		\path[->] (8) edge (12);
		\path[->] (9) edge (13);
		\path[->] (10) edge (14);
		\path[->] (11) edge (15);
		\path[->, bend right] (12) edge (16);
		\path[->, bend right] (13) edge (17);
		\path[->, bend right] (14) edge (18);
		\path[->, bend right] (15) edge (19);
		\path[->, bend right] (12) edge (0);
		\path[->, bend right] (13) edge (1);
		\path[->, bend right] (14) edge (2);
		\path[->, bend right] (15) edge (3);
		\path[->] (16) edge (22);
		\path[->] (16) edge (20);
		\path[->] (17) edge (21);
		\path[->] (19) edge (23);
		\path[->] (20) edge (27);
		\path[->] (20) edge (24);
		\path[->] (21) edge (25);
		\path[->] (22) edge (26);


                \end{tikzpicture}
            \end{document}
